api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java

```diff
@@ -717,56 +748,80 @@ private Optional<String> buildConceptIdSqlInClause(
+                .collect(Collectors.toList());
+        // find all answers for the questions
+        dbCriteriaAnswerIds = findPFHHSurveyAnswerIds(pfhhSurveyQuestionIds);
+      }
+    } else {
+      dbConceptSetConceptIds = findMultipleDomainConceptIds(domain, dbConceptSetIds);
+    }
+
+    if (dbConceptSetConceptIds.isEmpty() && dbCriteriaAnswerIds.isEmpty()) {
+      return Optional.empty();
+    } else {
+      StringBuilder queryBuilder = new StringBuilder();
+      Map<Boolean, List<DbConceptSetConceptId>> partitionSourceAndStandard =
+          dbConceptSetConceptIds.stream()
+              .collect(Collectors.partitioningBy(DbConceptSetConceptId::getStandard));
+      String standardConceptIds =
+          partitionSourceAndStandard.get(true).stream()
+              .map(c -> c.getConceptId().toString())
+              .collect(Collectors.joining(", "));
+      String sourceConceptIds =
+          partitionSourceAndStandard.get(false).stream()
+              .map(c -> c.getConceptId().toString())
+              .collect(Collectors.joining(", "));
+      if (!standardConceptIds.isEmpty()) {
+        queryBuilder.append(
+            BigQueryDataSetTableInfo.getConceptIdIn(domain, true)
+                .replaceAll("unnest", "")
+                .replaceAll("(@standardConceptIds)", standardConceptIds));
+      }
+      if (!sourceConceptIds.isEmpty()) {
         if (!standardConceptIds.isEmpty()) {
-          queryBuilder.append(
-              BigQueryDataSetTableInfo.getConceptIdIn(domain, true)
-                  .replaceAll("unnest", "")
-                  .replaceAll("(@standardConceptIds)", standardConceptIds));
+          queryBuilder.append(" OR ");
         }
-        if (!sourceConceptIds.isEmpty()) {
-          if (!standardConceptIds.isEmpty()) {
-            queryBuilder.append(" OR ");
-          }
-          queryBuilder.append(
-              BigQueryDataSetTableInfo.getConceptIdIn(domain, false)
-                  .replaceAll("unnest", "")
-                  .replaceAll("(@sourceConceptIds)", sourceConceptIds));
+        queryBuilder.append(
+            BigQueryDataSetTableInfo.getConceptIdIn(domain, false)
+                .replaceAll("unnest", "")
+                .replaceAll("(@sourceConceptIds)", sourceConceptIds));
+      }
+      if (Domain.SURVEY.equals(domain) && !dbCriteriaAnswerIds.isEmpty()) {
+        String answerConceptIds =
+            dbCriteriaAnswerIds.stream().map(Object::toString).collect(Collectors.joining(","));
+        if (queryBuilder.toString().contains("question_concept_id IN (")) {
+          queryBuilder.append(" OR ");
         }
-        return Optional.of("(" + queryBuilder + ")");
+        queryBuilder.append(
+            "answer_concept_id IN (@answerConceptIds)"
+                .replaceAll("@answerConceptIds", answerConceptIds));
       }
+      return Optional.of("(" + queryBuilder + ")");
     }
   }
```

Comment:
L#745-820 private Optional<String> buildConceptIdSqlInClause needs some refactor very hard to follow... just a suggestion- 75 lines of different if-else blocks...

Reply:
Refactor is not a priority right now. Please create a ticket for a refactor if you think you can make this better.